include/hip/hcc_detail/device_library_decls.h

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright (c) 2015 - present Advanced Micro Devices, Inc. All rights reserved.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+/**
+ *  @file  hcc_detail/device_library_decls.h
+ *  @brief Contains declarations for types and functions in device library.
+ */
+
+#ifndef HIP_INCLUDE_HIP_HCC_DETAIL_DEVICE_LIBRARY_DECLS_H
+#define HIP_INCLUDE_HIP_HCC_DETAIL_DEVICE_LIBRARY_DECLS_H
+
+#include "hip/hcc_detail/host_defines.h"
+
+extern "C" __device__ int32_t __ockl_activelane_u32(void);
+
+extern "C" __device__ uint __ockl_mul24_u32(uint, uint);
+extern "C" __device__ int __ockl_mul24_i32(int, int);
+extern "C" __device__ uint __ockl_mul_hi_u32(uint, uint);
+extern "C" __device__ int __ockl_mul_hi_i32(int, int);
+extern "C" __device__ uint __ockl_sad_u32(uint, uint, uint);
+
+extern "C" __device__ float __ocml_floor_f32(float);
+extern "C" __device__ float __ocml_rint_f32(float);
+extern "C" __device__ float __ocml_ceil_f32(float);
+extern "C" __device__ float __ocml_trunc_f32(float);
+
+// Introduce local address space
+#define __local __attribute__((address_space(3)))
+__device__ inline static __local char* __to_local(unsigned x) { return (__local char*)x; }
```

Comment:
HCC throws the warning:
```warning: cast to '__attribute__((address_space(3))) char *' from smaller integer type 'unsigned int' [-Wint-to-pointer-cast]```

Reply:
Should I hide this warning by adding CPP flag -Wno-int-to-pointer-cast ?

Reply:
The signature of this function doesn't look right to me.  I would expect `x` to be a generic pointer which is 64-bit.  Is there a unit test for this

Reply:
This is a workaround for a compiler bug. It relies on compiler inserting an addrspacecast when calling this function.

Aaron, since this relies on a compiler workaround which is not in amd-common branch, there is no point of putting this change into HIP. Can you separate this change (and those depending on it) and keep it in hip-clang branch only? Then you can update this pull request without this part of the change.

Reply:
sorry I mistaken this function with another function.

This is not the OpenCL __to_local function. It is a static function which is called by  __get_dynamicgroupbaseptr.

Reply:
The warning comes from host compile stage. Although there is no IR emitted for this device function in host compilation stage, it still needs to be parsed. Then a pointer in addrspace 3 is 64 bit for host, which causes the warning. Since this function is valid only for device compilation, we should guard it with __HIP__DEVICE_COMPILE.

Reply:
I will guard the original __to_local with `__HIP_DEVICE_COMPILE__` since a local pointer should be the same size as uint32_t on device side.

Reply:
Seems if I surround __to_local with that guard, I will need to surround get_dynamicgroupbaseptr caller functions as well.